module-system/sov-modules-api/src/containers/kernel_value.rs

```diff
@@ -0,0 +1,396 @@
+                self.prefix(),
+                key,
+                StateMapAccessor::<u64, V, Codec, KernelWorkingSet<'a, C>>::codec(self),
+            )
+        }
+
+        fn remove_or_err<Q>(
+            &self,
+            key: &Q,
+            working_set: &mut KernelWorkingSet<'a, C>,
+        ) -> Result<V, crate::StateMapError>
+        where
+            Codec: StateCodec,
+            <Codec as StateCodec>::KeyCodec: sov_modules_core::EncodeKeyLike<Q, u64>,
+            <Codec as StateCodec>::ValueCodec: StateValueCodec<V>,
+            Q: ?Sized,
+        {
+            self.remove(key, working_set).ok_or_else(|| {
+                crate::StateMapError::MissingValue(
+                    self.prefix().clone(),
+                    sov_modules_core::StorageKey::new(
+                        self.prefix(),
+                        key,
+                        StateMapAccessor::<u64, V, Codec, KernelWorkingSet<'a, C>>::codec(self)
+                            .key_codec(),
+                    ),
+                )
+            })
+        }
+
+        fn delete<Q>(&self, key: &Q, working_set: &mut KernelWorkingSet<'a, C>)
+        where
+            Codec: StateCodec,
+            <Codec as StateCodec>::KeyCodec: sov_modules_core::EncodeKeyLike<Q, u64>,
+            Q: ?Sized,
+        {
+            working_set.delete_value(
+                self.prefix(),
+                key,
+                StateMapAccessor::<u64, V, Codec, KernelWorkingSet<'a, C>>::codec(self),
+            );
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use sov_mock_da::MockDaSpec;
+    use sov_modules_core::capabilities::mocks::MockKernel;
+    use sov_modules_core::{Address, Context, KernelWorkingSet, Prefix, WorkingSet};
+    use sov_state::{DefaultStorageSpec, ProverStorage};
+
+    use crate::default_context::DefaultContext;
+    use crate::KernelStateValue;
+    #[test]
+    fn test_kernel_state_value_as_value() {
+        use crate::StateValueAccessor;
+        let tmpdir = tempfile::tempdir().unwrap();
+        let storage = ProverStorage::<DefaultStorageSpec>::with_path(tmpdir.path()).unwrap();
+        let mut working_set: WorkingSet<DefaultContext> = WorkingSet::new(storage);
+
+        let prefix = Prefix::new(b"test".to_vec());
+        let value = KernelStateValue::<u64>::new(prefix.clone());
+
+        // Initialize a value in the kernel state during slot 4
+        {
+            let kernel = MockKernel::<DefaultContext, MockDaSpec>::new(4, 1);
+            let mut kernel_state = KernelWorkingSet::from_kernel(&kernel, &mut working_set);
+            value.set(&100, &mut kernel_state);
+            assert_eq!(value.get(&mut kernel_state), Some(100));
+        }
+
+        {
+            {
+                let mut versioned_state =
+                    working_set.versioned_state(&DefaultContext::new(Address::from([1; 32]), 1));
+                // Try to read the value from user space with the slot number set to 1. Should fail.
```

Comment:
Question:

Where does the check of `VersionedWorkingSet` not returning values until they're in the past happen? Couldn't find it.

Reply:
Chatted offline. The "check" is internal to the prefix generation logic